be2-scala/src/main/scala/ch/epfl/pop/decentralized/GossipManager.scala

```diff
@@ -0,0 +1,123 @@
+import ch.epfl.pop.model.network.method.{GreetServer, Rumor}
+import ch.epfl.pop.model.network.{JsonRpcRequest, JsonRpcResponse, MethodType}
+import ch.epfl.pop.pubsub.AskPatternConstants
+import ch.epfl.pop.pubsub.ClientActor.ClientAnswer
+import ch.epfl.pop.pubsub.graph.{ErrorCodes, GraphMessage, PipelineError}
+import ch.epfl.pop.storage.DbActor
+import ch.epfl.pop.storage.DbActor.DbActorReadRumor
+
+import scala.concurrent.Await
+import scala.util.Random
+
+final case class GossipManager(
+    dbActorRef: AskableActorRef,
+    monitorRef: ActorRef,
+    connectionMediator: AskableActorRef,
+    stopProbability: Double = 0.5
+) extends Actor with AskPatternConstants {
+
+  private type ServerInfos = (ActorRef, GreetServer)
+
+  monitorRef ! GossipManager.Ping()
+  connectionMediator ? ConnectionMediator.Ping()
+
+  private var activeGossipProtocol: Map[JsonRpcRequest, List[ServerInfos]] = Map.empty
+
+  private def isRumorNew(rumor: Rumor): Boolean = {
+    val readRumorDb = dbActorRef ? DbActor.ReadRumor(rumor.senderPk -> rumor.rumorId)
+    Await.result(readRumorDb, duration) match
+      case DbActorReadRumor(foundRumors) =>
+        foundRumors match
+          case Some(_) => false
+          case None    => true
+  }
+
+  private def getPeersForRumor(jsonRpcRequest: JsonRpcRequest): List[ServerInfos] = {
+    val activeGossip = activeGossipProtocol.get(jsonRpcRequest)
+    activeGossip match
+      case Some(peersInfosList) => peersInfosList
+      case None                 => List.empty
+  }
+
+  private def sendRumorToRandomPeer(rumorRpc: JsonRpcRequest): Unit = {
+    // checks the peers to which we already forwarded the message
+    val activeGossip: List[ServerInfos] = getPeersForRumor(rumorRpc)
+    // selects a random peer from remaining peers
+    val randomPeer = connectionMediator ? ConnectionMediator.GetRandomPeer(activeGossip.map(_._1))
+    Await.result(randomPeer, duration) match {
+      // updates the list based on response
+      // if some peers are available we send
+      case ConnectionMediator.GetRandomPeerAck(serverRef, greetServer) =>
+        val alreadySent: List[ServerInfos] = activeGossip :+ (serverRef -> greetServer)
+        activeGossipProtocol += (rumorRpc -> alreadySent)
+        serverRef ! ClientAnswer(
+          Right(rumorRpc)
+        )
+      // else remove entry
+      case ConnectionMediator.NoPeer =>
+        activeGossipProtocol = activeGossipProtocol.removed(rumorRpc)
+    }
+  }
+
+  private def processResponse(response: JsonRpcResponse): Unit = {
+    val activeGossipPeers = activeGossipProtocol.filter((k, _) => k.id == response.id)
+    // response is expected because only one entry exists
+    if (activeGossipPeers.size == 1) {
+      activeGossipPeers.foreach { (rumorRpc, _) =>
+        if (response.result.isEmpty && Random.nextDouble() < stopProbability) {
+          activeGossipProtocol -= rumorRpc
+        } else {
+          sendRumorToRandomPeer(rumorRpc)
+        }
+      }
```

Comment:
Same as discussed in meeting, shoudln't we handle case where size != 1?

Reply:
I am not sure what could be done in that case. For two reasons : 
- If there is 0 match => response is related to a gossip that isn't active, so nothing should be done related to this response
- If there is 2 matches it means that we have two rumor to which this response could be related to, and for now we don't have a way to distinguish them.

What do you see as the most suitable behaviour ? Select a random rumor from the pool of possible rumors and process ? or ignore  ? (or any other suggestion ?)